CRM/Core/BAO/ConfigSetting.php

```diff
@@ -147,14 +147,20 @@ public static function applyLocale($settings, $activatedLocales) {
     // Grab session reference.
     $session = CRM_Core_Session::singleton();
 
-    // Set flag for multi-language setup.
-    $multiLang = (bool) $activatedLocales;
-
     // Initialise the default and chosen locales.
     $defaultLocale = $settings->get('lcMessages');
     $chosenLocale = NULL;
 
+    // Parse multi lang locales
+    $multiLangLocales = $activatedLocales ? explode(CRM_Core_DAO::VALUE_SEPARATOR, $activatedLocales) : NULL;
+
+    // On multilang, defaultLocale should be one of the activated locales
+    if ($multiLangLocales && !in_array($defaultLocale, $multiLangLocales)) {
+      $defaultLocale = NULL;
+    }
+
     // When there is a choice of permitted languages.
+    // Q: why would this be different from the locales?
```

Comment:
(*Tangential subthread about the comment*)

Yeah, that kind of question hurts the brain. [The general problem](https://lab.civicrm.org/dev/translation/-/issues/78) is [this](https://github.com/civicrm/civicrm-core/blob/b37ca98db9a6f055f3f504a77651f9a3d2c36d6c/Civi/Core/Locale.php#L19-L47):

> The software comes with a list of 200 communication-locales (OptionValues), and admins may register more. There are only ~50 locales supported by `ts()` and 1-3 locales in the DB (for a typical business-entity). If you use any of these other locales, then some services must raise errors (or fallback to an alternate locale).

Those numbers are completely different. The situation invites edge-cases. If you operate in these circumstances, then the system has some *de-facto negotiation behavior*. ("We see the user prefers `xx_XX`, but we cannot meet it head-on, so we switch to resource `yy_YY` or throw error `Zzz`.")

Civi has two implementations of negotiation behavior. (*Naturally! Why wouldn't it?*)

* `applyLocale()` identifies the *preferred* locale by examining `$_REQUEST` data and CMS/user data...  which is good.  Then it decides whether/how it can satisfy the preference, but it doesn't try very hard. (It always need a full match in all layers. It doesn't support the `partial_locales` setting.)
* `setLocale()` (aka `Locale::negotiate()`) requires you to tell it the `$preferred` locale. But it [works harder](https://github.com/civicrm/civicrm-core/blob/b37ca98db9a6f055f3f504a77651f9a3d2c36d6c/Civi/Core/Locale.php#L176-L217) to satisfy the user's preference. (It does respect `partial_locales`.)

Perhaps... in an ideal world... we would put these together (eg `applyLocale()` would call `setLocale($preferred)`). But validating that change is another adventure...

Reply:
Thanks @totten . Have dropped a reference to this into the comment.

Anything non-tangential outstanding? :slightly_smiling_face: